src/date/date.rs

```diff
@@ -0,0 +1,283 @@
+    Date,
+    Seconds,
+    Ns,
+}
+
+impl<'a> From<&'a str> for Rfc3339Format {
+    fn from(s: &str) -> Self {
+        match s {
+            DATE => Rfc3339Format::Date,
+            SECONDS => Rfc3339Format::Seconds,
+            NS => Rfc3339Format::Ns,
+            // Should be caught by clap
+            _ => panic!("Invalid format: {}", s),
+        }
+    }
+}
+
+pub fn uumain(args: Vec<String>) -> i32 {
+
+    let settings = parse_cli(args);
+
+    if let Some(_time) = settings.set_to {
+        unimplemented!();
+        // Probably need to use this syscall:
+        // https://doc.rust-lang.org/libc/i686-unknown-linux-gnu/libc/fn.clock_settime.html
+
+    } else {
+        // Declare a file here because it needs to outlive the `dates` iterator.
+        let file: File;
+
+        // Get the current time, either in the local time zone or UTC.
+        let now: DateTime<FixedOffset> = match settings.utc {
+            true => {
+                let now = UTC::now();
+                now.with_timezone(&now.offset().fix())
+            }
+            false => {
+                let now = Local::now();
+                now.with_timezone(now.offset())
+            }
+        };
+
+        /// Parse a `String` into a `DateTime`.
+        /// If it fails, return a tuple of the `String` along with its `ParseError`.
+        fn parse_date(s: String)
+                      -> Result<DateTime<FixedOffset>, (String, chrono::format::ParseError)> {
+            // TODO: The GNU date command can parse a wide variety of inputs.
+            s.parse().map_err(|e| (s, e))
+        }
+
+        // Iterate over all dates - whether it's a single date or a file.
+        let dates: Box<Iterator<Item = _>> = match settings.date_source {
+            DateSource::Custom(ref input) => {
+                let date = parse_date(input.clone());
+                let iter = std::iter::once(date);
+                Box::new(iter)
+            }
+            DateSource::File(ref path) => {
+                file = File::open(path).unwrap();
+                let lines = BufReader::new(file).lines();
+                let iter = lines.filter_map(Result::ok).map(parse_date);
+                Box::new(iter)
+            }
+            DateSource::Now => {
+                let iter = std::iter::once(Ok(now));
+                Box::new(iter)
+            }
+        };
+
+        let format_string = make_format_string(&settings);
+
+        // Format all the dates
+        for date in dates {
+            match date {
+                Ok(date) => {
+                    let formatted = date.format(format_string);
+                    println!("{}", formatted);
+                }
+                Err((input, _err)) => {
+                    println!("date: invalid date '{}'", input);
+                }
+            }
+        }
+    }
+
+    0
+}
+
+
+/// Handle command line arguments.
+fn parse_cli(args: Vec<String>) -> Settings {
+    let matches = clap_app!(
```

Comment:
I love clap, but we tend to use a convention around getopt because for several utilities clap is incapable of correctly reproducing GNU coreutils behavior. It's not something that needs to be corrected now, just a note for the future.